woocommerce/woocommerce-ios
# Order Stats: Custom date range tab

Implements https://github.com/woocommerce/woocommerce-ios/issues/842

This adds a new tab on the store dashboard so users can specify a custom date range for the orders and visits stats.

High level sub tasks:
- [x] Update models to fetch and query order stats with a custom range
- [x] Add modal controller for selecting a custom range
  - [x] Allow selection range from/to date  
  - [x] ~~Allow changing granularity~~ Auto select granularity based on the date range
  - [x] Persist custom range
- [x] Display selected range under the new custom tab

## Before

![image](https://user-images.githubusercontent.com/57791/56346334-28f1dd80-61c2-11e9-8650-bc1124bdb2a6.png)
![IMG_0064](https://user-images.githubusercontent.com/57791/56353485-c6a0d900-61d1-11e9-9519-314dbe19e31c.PNG)

## After

![custom-range-demo](https://user-images.githubusercontent.com/57791/56353092-ebe11780-61d0-11e9-957b-dd72c73c494b.gif)
![custom-range-demo-ipad](https://user-images.githubusercontent.com/57791/56362540-6c604200-61ea-11e9-9b7f-bbde42f4beb2.gif)

## Changes

- Updated Networking, Storage and Yosemite to pass a new `queryID` string parameter for orders and visits stats to be able to isolate stats results based on it. Before my changes, the `granularity` parameter was used to achieve that. But with the addition of the custom tab, this wasn't enough anymore as it needed to fetch and store stats with different granularities, hence conflicting with the existing tabs data.
- Updated existing tests and added new tests to confirm this new isolation based on `queryID` works as expected.
- Added new tab in `StoreStatsViewController`
- Added `CustomDateRangeSelectionViewController` which contain the `UIDatePicker` based range selection similar to what's done in the Apple Calendar app
- Updated `PeriodDataViewController` to support the new custom range tab case. It shows the custom range under the tab. Tapping on it opens `CustomDateRangeSelectionViewController`
- Added `StatsQueryOptions` to group the parameters to query/fetch the stats depending on the 2 different use cases: 
  - latest date with granularity
  - date range with automatic granularity
- Updated analytics events properties for `dashboard_main_stats_date` and `dashboard_main_stats_loaded` to take the custom range into account
- Added new analytics events `dashboard_main_stats_custom_range_edit_tapped` and `dashboard_main_stats_custom_range_applied`

## Testing

- Checkout the branch, run the tests
- Open the app, 
- Select the custom tab
  - Expected behavior:
    - when the app is first launched, it fetches and displays the last 14 days of data
    - when the app is relaunched, it displays existing cached data while fetching new data
    - when pull-to-refresh is used, it fetches new data and updates the graph
- Tap `edit` next to the displayed range under the custom tab
  - Expected behavior:
    - A modal controller is presented to select the custom date range
      - select a new range
      - tap apply
         - when the the start date is BEFORE the end date, it closes the modal and fetches the data for the new range
           - if the range duration is > 365 days, it fetches the data with `year` granularity.
           - if the range duration is > 31 * 6 days, it fetches the data with `month` granularity.
           - if the range duration is > 31 days, it fetches the data with `week` granularity.
           - otherwise it fetches the data with `day` granularity
         - when the the start date is AFTER the end date, it displays an alert telling the start date must be before the end date
      - relaunch the app, the previously selected range should be used

## Known limitations and future work

- The custom date range is using `UserDefaults` for persistence for simplicity. It could benefit from the work done by @ctarda on https://github.com/woocommerce/woocommerce-ios/pull/883 to support local site specific settings.
- The stats API only allows querying using `date` (latest date to include), `unit` (granularity: `day`, `week`, `month`, `year`), `quantity` (how many units, doesn't support decimals). So when choosing a range, the current code maps it to the appropriate `date`, `granularity` and `quantity`, which means it returns data for more than the selected range. For instance if you select a range of 1.5 years with `year` granularity, you'll have to use 2 as the `quantity` and hence you'll get data for 2 years till the selected end date.
- There's an ordering issue in the chart revealed by the new use case added in the PR. I've opened issue https://github.com/woocommerce/woocommerce-ios/issues/887 with the details.
- Whenever you close the date range modal or settings, there's a slight visual glitch, the bar graph appear fully for a brief moment before animating. This bug was already present before my changes. I've opened https://github.com/woocommerce/woocommerce-ios/issues/886 to track this.

Update release notes:
